src/app/codegen-data-model-provider/CodegenDataModelProvider.cpp

```diff
@@ -396,21 +412,47 @@ bool CodegenDataModelProvider::EndpointExists(EndpointId endpoint)
     return (emberAfIndexFromEndpoint(endpoint) != kEmberInvalidEndpointIndex);
 }
 
-EndpointId CodegenDataModelProvider::FirstEndpoint()
+std::optional<DataModel::EndpointInfo> CodegenDataModelProvider::GetEndpointInfo(EndpointId endpoint)
+{
+    uint16_t endpoint_idx = emberAfIndexFromEndpoint(endpoint);
+    if (endpoint_idx != kEmberInvalidEndpointIndex && emberAfEndpointIndexIsEnabled(endpoint_idx))
+    {
+        DataModel::EndpointInfo info(emberAfParentEndpointFromIndex(endpoint_idx));
+        if (IsFlatCompositionForEndpoint(endpoint))
+        {
+            info.compositionPattern = DataModel::EndpointCompositionPattern::kFullFamilyPattern;
+        }
+        else if (IsTreeCompositionForEndpoint(endpoint))
+        {
+            info.compositionPattern = DataModel::EndpointCompositionPattern::kTreePattern;
+        }
+        return std::make_optional(info);
+    }
+    return std::nullopt;
+}
+
+DataModel::EndpointEntry CodegenDataModelProvider::FirstEndpoint()
 {
     // find the first enabled index
     const uint16_t lastEndpointIndex = emberAfEndpointCount();
     for (uint16_t endpoint_idx = 0; endpoint_idx < lastEndpointIndex; endpoint_idx++)
     {
         if (emberAfEndpointIndexIsEnabled(endpoint_idx))
         {
-            mEndpointIterationHint = endpoint_idx;
-            return emberAfEndpointFromIndex(endpoint_idx);
+            mEndpointIterationHint                 = endpoint_idx;
+            DataModel::EndpointEntry endpointEntry = DataModel::EndpointEntry::kInvalid;
+            endpointEntry.id                       = emberAfEndpointFromIndex(endpoint_idx);
+            auto endpointInfo                      = GetEndpointInfo(endpointEntry.id);
```

Comment:
Passing the id to GetEndpointInfo sort of defeats the purpose of having mEndpointIterationHint, no?  Because it does another linear walk.....

Reply:
Yes, I added a new `GetEndpointInfoAtIndex()` function and call it here.